openshift/WFLY-14936_New_S2I_Image.adoc

```diff
@@ -0,0 +1,228 @@
+N.B.: In case of openshift-launch.sh (legacy entry point), no configuration is applied. Configuration of the server is delegated 
+to the ```openshift-launch.sh``` script that handles it all. The env variable ```SERVER_LAUNCH_SCRIPT_OVERRIDE``` can be used to override the server launch script file name. 
+The file must be located in ```$JBOSS_HOME/bin``` directory.
+** *WARNING. This needs to be reviewed at the light of future evolutions in the transaction subsystem.* Computes the value of the jboss.node.name system property (required by transactions). By default a truncated to 23 characters name based on the hostname is computed. This can be 
+overridden by using the ```JBOSS_NODE_NAME``` env variable.
+** Computes the JVM options to be conveyed to the server
+** Appends the content of ```JAVA_OPTS_APPEND``` env variable to ```JAVA_OPTS```.
+** If the env variable ```CLI_LAUNCH_SCRIPT=<cli script path>``` has been set, the referenced CLI script is executed during boot. The file can be an absolute path or a path relative to ```JBOSS_HOME``` directory.
+** Handles POD termination with CLI shutdown. Can be disabled with ```CLI_GRACEFUL_SHUTDOWN=true``` env variable. N.B.: 
+the ```PORT_OFFSET``` env variable is to be used to convey a port offset, This allows the CLI shutdown logic to retrieve the port offset when interacting with the server.
+** Set default value for management (0.0.0.0) and public (value of ```hostname -i``` command) interfaces. 
+    The default values can be specified with ```SERVER_PUBLIC_BIND_ADDRESS``` and  ```SERVER_MANAGEMENT_BIND_ADDRESS```.
+** Enables the server statistics. Can be overridden with ```SERVER_ENABLE_STATISTICS=false```.
+** Calls the server launch script passing it the ```SERVER_ARGS``` env variable that can contain extra arguments to provide to the server.
+
+The API to configure the server launch (for both the builder and runtime images entry-point) is defined in https://github.com/wildfly/wildfly-cekit-modules/blob/v2/jboss/container/wildfly/run/api/module.yaml[this cekit module]
+
+==== WildFly runtime image
+
+* The runtime image is built from openjdk JRE 11 image.
+* Has the same versioning that the S2I Builder image.
+* It is deployed on quay.io: ```quay.io/wildfly/wildfly-runtime-jre11```
+* It contains the user jboss
+* Image produced from this runtime image is runnable.
+* It contains the same entry-point present in the builder image.
+* Examples of Docker file to install a server inside the runtime image:
+
+Local WildFly installation copied to the image
+```
+FROM quay.io/wildfly/wildfly-runtime-jre11:latest
+COPY --chown=jboss:root vanilla-wildfly $JBOSS_HOME
+RUN chmod -R ug+rwX $JBOSS_HOME
+```
+
+Chained build from builder image
+```
+FROM quay.io/wildfly/wildfly-runtime-jre11:latest
+COPY --from=builder --chown=jboss:root /opt/wildfly $JBOSS_HOME
+RUN chmod -R ug+rwX $JBOSS_HOME
+```
+
+==== WildFly Cloud legacy Galleon feature-pack
```

Comment:
@jfdenise Imagine I have existing s2i applications relying on environment properties. How exactly should look like migration to this new way? Is there somewhere any example I can have a look? With my experience from bootable jar approach I assume I will have to adjust pom.xml of application to build server using maven plugin. In that pom.xml I will have feature pack for wildfly server version which I want to use and also `wildfly-cloud-legacy-galleon-pack` as I want to be backward compatible and use environment properties as I was used to.

If my assumptions are right, are those app changes applicable by PM as there is big focus on easy migration between versions. Ideally without app changes. Or is it something which can be covered by migration tool?

Reply:
@mchoma , yes, what you describe is what a user would need to do. You can see an example: https://github.com/wildfly/wildfly-s2i/blob/v2/test/test-app-clustering/pom.xml
Then behave tests that uses env variables: https://github.com/wildfly/wildfly-s2i/blob/v2/wildfly-builder-image/tests/features/jgroups.feature#L4
This does require a change in application pom.xml.